### ISSUE TITLE
ActiveRecord::StatementInvalid in Thredded::MessageboardsController#index

### ISSUE THREAD

**Issue by toommz** [thredded/thredded#822]
Hi,

On a Rails 6 rc2 project, when attempting to access forum index, I'm encountering the following error :

`PG::UndefinedTable: ERROR: missing FROM-clause entry for table "thredded_topics" `

The generated SQL is:

> "SELECT \"thredded_messageboards\".* FROM \"thredded_messageboards\" LEFT OUTER JOIN \"thredded_user_topic_read_states\" ON \"thredded_messageboards\".\"id\" = \"thredded_user_topic_read_states\".\"messageboard_id\" AND \"thredded_user_topic_read_states\".\"postable_id\" = \"thredded_topics\".\"id\" AND \"thredded_user_topic_read_states\".\"user_id\" IS NULL AND \"thredded_user_topic_read_states\".\"unread_posts_count\" = 0 INNER JOIN \"thredded_topics\" ON \"thredded_topics\".\"messageboard_id\" = \"thredded_messageboards\".\"id\" WHERE (\"thredded_topics\".\"moderation_state\" != 2 OR \"thredded_topics\".\"user_id\" IS NULL) GROUP BY \"thredded_messageboards\".\"id\""

I tried to find a solution but quickly got lost with all those arel tables in [this class method](https://github.com/thredded/thredded/blob/master/app/models/thredded/messageboard.rb#L127)

**Comment by glebm** [comment]
Hmm, the generated SQL looks OK to me.

**Comment by toommz** [comment]
PostgreSQL is complaining about this SQL, but this one is fine:

`SELECT "thredded_messageboards".* FROM "thredded_messageboards" INNER JOIN "thredded_topics" ON "thredded_topics"."messageboard_id" = "thredded_messageboards"."id" LEFT OUTER JOIN "thredded_user_topic_read_states" ON "thredded_messageboards"."id" = "thredded_user_topic_read_states"."messageboard_id" AND "thredded_user_topic_read_states"."postable_id" = "thredded_topics"."id" AND "thredded_user_topic_read_states"."user_id" IS NULL AND "thredded_user_topic_read_states"."unread_posts_count" = 0 WHERE ("thredded_topics"."moderation_state" != 2 OR "thredded_topics"."user_id" IS NULL) GROUP BY "thredded_messageboards"."id"`

I just switched the INNER JOIN for thredded_topics before referencing it in the LEFT OUTER JOIN. But as I sais earlier, Arel is kinda black magic to me.

Here's the #to_sql result on a Rails 5 project:

`"SELECT \"thredded_messageboards\".* FROM \"thredded_messageboards\" INNER JOIN \"thredded_topics\" ON \"thredded_topics\".\"messageboard_id\" = \"thredded_messageboards\".\"id\" LEFT OUTER JOIN \"thredded_user_topic_read_states\" ON \"thredded_messageboards\".\"id\" = \"thredded_user_topic_read_states\".\"messageboard_id\" AND \"thredded_user_topic_read_states\".\"postable_id\" = \"thredded_topics\".\"id\" AND \"thredded_user_topic_read_states\".\"user_id\" = 1 AND \"thredded_user_topic_read_states\".\"unread_posts_count\" = 0 WHERE (\"thredded_topics\".\"moderation_state\" != 2 OR \"thredded_topics\".\"user_id\" = 1) GROUP BY \"thredded_messageboards\".\"id\""`

**Comment by toommz** [comment]
Ok so playing with black magic may be dangerous, but this one is working:

```ruby
def unread_topics_counts(user:, topics_scope: Thredded::Topic.all)
  messageboards = arel_table
  read_states = Thredded::UserTopicReadState.arel_table
  topics = topics_scope.arel_table
  joins(
    messageboards.join(topics).on(messageboards[:id].eq(topics[:messageboard_id]))
    .outer_join(read_states).on(
      messageboards[:id].eq(read_states[:messageboard_id])
        .and(read_states[:postable_id].eq(topics[:id]))
        .and(read_states[:user_id].eq(user.id))
        .and(read_states[:unread_posts_count].eq(0))
    ).join_sources
  ).merge(topics_scope).group(messageboards[:id]).pluck(
    :id,
    Arel::Nodes::Subtraction.new(topics[:id].count, read_states[:id].count)
  ).to_h
end
```

The resulting SQL (as always, before the #pluck):

`"SELECT \"thredded_messageboards\".* FROM \"thredded_messageboards\" INNER JOIN \"thredded_topics\" ON \"thredded_messageboards\".\"id\" = \"thredded_topics\".\"messageboard_id\" LEFT OUTER JOIN \"thredded_user_topic_read_states\" ON \"thredded_messageboards\".\"id\" = \"thredded_user_topic_read_states\".\"messageboard_id\" AND \"thredded_user_topic_read_states\".\"postable_id\" = \"thredded_topics\".\"id\" AND \"thredded_user_topic_read_states\".\"user_id\" = 1 AND \"thredded_user_topic_read_states\".\"unread_posts_count\" = 0 WHERE (\"thredded_topics\".\"moderation_state\" != 2 OR \"thredded_topics\".\"user_id\" = 1) GROUP BY \"thredded_messageboards\".\"id\""`

**Comment by glebm** [comment]
So the SQL is missing a join with topics, even though our query starts with `joins(:topics)`.

This sounds like a bug in Rails 6 rc2.

**Comment by glebm** [comment]
Filed a bug in Rails. Best revert to RC1 for now, who knows what else is broken.

**Comment by glebm** [comment]
Discussion in the Rails bug shed some light on this and I was able to come up with a work-around. Hopefully this can be fixed in Rails before 6.0.0 final so that the work-around can be removed in the future.

**Comment by toommz** [comment]
Thanks for taking the time on this. Ironically I wanted to had a rails_gte_6 but thought you’ll find it a bit sluggish :)

I agree with you about the perceived publicness of Arel, didn’t even know the api was private.

**Comment by glebm** [comment]
I've released v0.16.12 with the workaround. This version of the gem restricts Rails to `<= 6.0.0.rc2`, just in case the bug is fixed in RC3 / final.
